README.md
# SMS Monolog Handler
[![Latest Version](https://img.shields.io/github/release/tylercd100/monolog-sms.svg?style=flat-square)](https://github.com/tylercd100/monolog-sms/releases)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE.md)
[![Build Status](https://travis-ci.org/tylercd100/monolog-sms.svg?branch=master)](https://travis-ci.org/tylercd100/monolog-sms)
[![Scrutinizer Code Quality](https://scrutinizer-ci.com/g/tylercd100/monolog-sms/badges/quality-score.png?b=master)](https://scrutinizer-ci.com/g/tylercd100/monolog-sms/?branch=master)
[![Code Coverage](https://scrutinizer-ci.com/g/tylercd100/monolog-sms/badges/coverage.png?b=master)](https://scrutinizer-ci.com/g/tylercd100/monolog-sms/?branch=master)
[![Dependency Status](https://www.versioneye.com/user/projects/56f3252c35630e0029db0187/badge.svg?style=flat)](https://www.versioneye.com/user/projects/56f3252c35630e0029db0187)
[![Total Downloads](https://img.shields.io/packagist/dt/tylercd100/monolog-sms.svg?style=flat-square)](https://packagist.org/packages/tylercd100/monolog-sms)

A Monolog Handler for SMS messaging services

Currently supported
- [Twilio](https://www.twilio.com/)
- [Clickatell](https://www.clickatell.com/)
- [Plivo](https://www.plivo.com/)

## Installation

Install via [composer](https://getcomposer.org/) - In the terminal:
```bash
composer require tylercd100/monolog-sms
```

## Usage
For Plivo:
```php
use Tylercd100\Monolog\Handler\PlivoHandler;

$handler = new PlivoHandler($token,$auth_id,$fromPhoneNumber,$toPhoneNumber);
$logger  = new Monolog\Logger('channel.name');
$logger->pushHandler($handler);
$logger->addCritical("Foo Bar!");
```

For Twilio:
```php
use Tylercd100\Monolog\Handler\TwilioHandler;

$handler = new TwilioHandler($secret,$sid,$fromPhoneNumber,$toPhoneNumber);
$logger  = new Monolog\Logger('channel.name');
$logger->pushHandler($handler);
$logger->addCritical("Foo Bar!");
```

For Clickatell:
```php
use Tylercd100\Monolog\Handler\ClickatellHandler;

$handler = new ClickatellHandler($authToken,$fromPhoneNumber (/*Optional*/),$toPhoneNumber (/*String|Array*/));
$logger  = new Monolog\Logger('channel.name');
$logger->pushHandler($handler);
$logger->addCritical("Foo Bar!");
```

CHANGELOG.md
# Changelog

All notable changes to `Monolog SMS` will be documented in this file.

### 2.0.0
- Add PHP 7.2, 7.3, 7.4 and 8.0 support
- Upgrade from Monolog 1.12 to 2.0
- Update PHPUnit tests
- Drop PHP 5.5, 5.6, 7.0, 7.1 and HHVM support

### 1.3.0
- Added support for Clickatell

### 1.2.0
- Added character limit

### 1.1.0
- Added support for Twilio

### 1.0.0
- Initial release


tests/TestCase.php
<?php

/*
 * This file is part of the Monolog package.
 *
 * (c) Jordi Boggiano <j.boggiano@seld.be>
 *
 * For the full copyright and license information, please view the LICENSE
 * file that was distributed with this source code.
 */

namespace Tylercd100\Monolog\Tests;

use ReflectionClass;
use Monolog\Logger;
use Monolog\Formatter\FormatterInterface;
use PHPUnit\Framework\TestCase as PHPUnitTestCase;

class TestCase extends PHPUnitTestCase
{
    /**
     * @return array Record
     */
    protected function getRecord($level = Logger::WARNING, $message = 'test', $context = array())
    {
        return array(
            'message' => $message,
            'context' => $context,
            'level' => $level,
            'level_name' => Logger::getLevelName($level),
            'channel' => 'test',
            'datetime' => \DateTime::createFromFormat('U.u', sprintf('%.6F', microtime(true))),
            'extra' => array(),
        );
    }

    /**
     * @return array
     */
    protected function getMultipleRecords()
    {
        return array(
            $this->getRecord(Logger::DEBUG, 'debug message 1'),
            $this->getRecord(Logger::DEBUG, 'debug message 2'),
            $this->getRecord(Logger::INFO, 'information'),
            $this->getRecord(Logger::WARNING, 'warning'),
            $this->getRecord(Logger::ERROR, 'error'),
        );
    }

    /**
     * @return Monolog\Formatter\FormatterInterface
     */
    protected function getIdentityFormatter()
    {
        $formatter = $this->getMockBuilder(FormatterInterface::class)->getMock();
        $formatter->expects($this->any())
            ->method('format')
            ->will(
                $this->returnCallback(function ($record) {
                    return $record['message'];
                })
            );

        return $formatter;
    }

    /**
     * Helper to return protected or private property value of an object
     *
     * @param  mixed  $object
     * @param  string $property
     * @return mixed
     */
    protected function accessProtected($object, $property)
    {
        $reflection = new ReflectionClass($object);
        $propertyReflection = $reflection->getProperty($property);
        $propertyReflection->setAccessible(true);

        return $propertyReflection->getValue($object);
    }

    /**
     * {@inheritdoc}
     */
    public static function assertMatchesRegularExpression(string $pattern, string $string, string $message = ''): void
    {
        if (method_exists(PHPUnitTestCase::class, 'assertMatchesRegularExpression')) {
            // For PHPUnit >= 9
            parent::assertMatchesRegularExpression($pattern, $string, $message);
        } else {
            // For PHPUnit < 9
            static::assertRegExp($pattern, $string, $message);
        }
    }
}


tests/PlivoHandlerTest.php
<?php

namespace Tylercd100\Monolog\Tests;

use Exception;
use PHPUnit_Framework_TestCase;
use Tylercd100\Monolog\Handler\PlivoHandler;
use Monolog\Logger;
use Tylercd100\Monolog\Tests\TestCase;

class PlivoHandlerTest extends TestCase
{
    private $res;
    /** @var  PlivoHandler */
    private $handler;

    public function testCanBeInstantiatedAndProvidesDefaultFormatter()
    {
        $handler = new PlivoHandler('token', 'auth_id', '+15555555555', '+16666666666');

        $this->assertInstanceOf('Tylercd100\\Monolog\\Formatter\\SMSFormatter', $handler->getFormatter());
        $this->assertEquals('token',        $this->accessProtected($handler, 'authToken'));
        $this->assertEquals('auth_id',      $this->accessProtected($handler, 'authId'));
        $this->assertEquals('+15555555555', $this->accessProtected($handler, 'fromNumber'));
        $this->assertEquals('+16666666666', $this->accessProtected($handler, 'toNumber'));
    }

    public function testItThrowsExceptionWhenUsingDifferentVersionOtherThanV1()
    {
        $this->expectException(Exception::class);
        $handler = new PlivoHandler('token', 'auth_id', '+15555555555', '+16666666666', Logger::CRITICAL, true, true, 'plivo.foo.bar', 'v2');
    }

    public function testWriteHeader()
    {
        $this->createHandler();
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'test1'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/POST \/v1\/Account\/auth_id\/Message\/ HTTP\/1.1\\r\\nHost: api.plivo.com\\r\\nAuthorization: Basic YXV0aF9pZDp0b2tlbg==\\r\\nContent-Type: application\/json\\r\\nContent-Length: \d{2,4}\\r\\n\\r\\n/', $content);

        return $content;
    }

    public function testWriteCustomHostHeader()
    {
        $this->createHandler('token', 'auth_id', '+15555555555', '+16666666666', Logger::CRITICAL, true, true, 'plivo.foo.bar');
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'test1'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/POST \/v1\/Account\/auth_id\/Message\/ HTTP\/1.1\\r\\nHost: plivo.foo.bar\\r\\nAuthorization: Basic YXV0aF9pZDp0b2tlbg==\\r\\nContent-Type: application\/json\\r\\nContent-Length: \d{2,4}\\r\\n\\r\\n/', $content);

        return $content;
    }

    /**
     * @depends testWriteHeader
     */
    public function testWriteContent($content)
    {
        $this->assertMatchesRegularExpression('/{"src":"\+15555555555","dst":"\+16666666666","text":"test1"}/', $content);
    }

    public function testWriteContentV1WithoutToAndFromNumbers()
    {
        $this->createHandler('token', 'auth_id', false, null, Logger::CRITICAL, true, true, 'plivo.foo.bar');
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'test1'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/{"src":false,"dst":null,"text":"test1"}/', $content);

        return $content;
    }

    /**
     * @depends testWriteCustomHostHeader
     */
    public function testWriteContentNotify($content)
    {
        $this->assertMatchesRegularExpression('/{"src":"\+15555555555","dst":"\+16666666666","text":"test1"}/', $content);
    }

    public function testWriteWithComplexMessage()
    {
        $this->createHandler();
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'Backup of database example finished in 16 minutes.'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/{"src":"\+15555555555","dst":"\+16666666666","text":"Backup of database example finished in 16 minutes\."}/', $content);
    }

    private function createHandler($authToken = 'token', $authId = 'auth_id', $fromNumber = '+15555555555', $toNumber = '+16666666666', $level = Logger::CRITICAL, $bubble = true, $useSSL = true, $host = 'api.plivo.com', $version = PlivoHandler::API_V1)
    {
        $constructorArgs = array($authToken, $authId, $fromNumber, $toNumber, Logger::DEBUG, true, true, $host, $version);
        $this->res = fopen('php://memory', 'a');
        $this->handler = $this->getMockBuilder(PlivoHandler::class)
            ->setMethods(['fsockopen', 'streamSetTimeout', 'closeSocket'])
            ->setConstructorArgs($constructorArgs)
            ->getMock();

        $reflectionProperty = new \ReflectionProperty('\Monolog\Handler\SocketHandler', 'connectionString');
        $reflectionProperty->setAccessible(true);
        $reflectionProperty->setValue($this->handler, 'localhost:1234');

        $this->handler->expects($this->any())
            ->method('fsockopen')
            ->will($this->returnValue($this->res));
        $this->handler->expects($this->any())
            ->method('streamSetTimeout')
            ->will($this->returnValue(true));
        $this->handler->expects($this->any())
            ->method('closeSocket')
            ->will($this->returnValue(true));

        $this->handler->setFormatter($this->getIdentityFormatter());
    }
}


tests/TwilioHandlerTest.php
<?php

namespace Tylercd100\Monolog\Tests;

use Exception;
use PHPUnit_Framework_TestCase;
use Tylercd100\Monolog\Handler\TwilioHandler;
use Monolog\Logger;
use Tylercd100\Monolog\Tests\TestCase;

class TwilioHandlerTest extends TestCase
{
    private $res;
    /** @var  TwilioHandler */
    private $handler;

    public function testCanBeInstantiatedAndProvidesDefaultFormatter()
    {
        $handler = new TwilioHandler('token', 'auth_id', '+15555555555', '+16666666666');

        $this->assertInstanceOf('Tylercd100\\Monolog\\Formatter\\SMSFormatter', $handler->getFormatter());
        $this->assertEquals('token',        $this->accessProtected($handler, 'authToken'));
        $this->assertEquals('auth_id',      $this->accessProtected($handler, 'authId'));
        $this->assertEquals('+15555555555', $this->accessProtected($handler, 'fromNumber'));
        $this->assertEquals('+16666666666', $this->accessProtected($handler, 'toNumber'));
    }

    public function testItThrowsExceptionWhenUsingDifferentVersionOtherThanV1()
    {
        $this->expectException(Exception::class);
        $handler = new TwilioHandler('token', 'auth_id', '+15555555555', '+16666666666', Logger::CRITICAL, true, true, 'twilio.foo.bar', 'v2');
    }

    public function testWriteHeader()
    {
        $this->createHandler();
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'test1'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/POST \/2010-04-01\/Accounts\/auth_id\/Messages\.json HTTP\/1.1\\r\\nHost: api.twilio.com\\r\\nAuthorization: Basic YXV0aF9pZDp0b2tlbg==\\r\\nContent-Type: application\/json\\r\\nContent-Length: \d{2,4}\\r\\n\\r\\n/', $content);

        return $content;
    }

    public function testWriteCustomHostHeader()
    {
        $this->createHandler('token', 'auth_id', '+15555555555', '+16666666666', Logger::CRITICAL, true, true, 'twilio.foo.bar');
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'test1'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/POST \/2010-04-01\/Accounts\/auth_id\/Messages.json HTTP\/1.1\\r\\nHost: twilio.foo.bar\\r\\nAuthorization: Basic YXV0aF9pZDp0b2tlbg==\\r\\nContent-Type: application\/json\\r\\nContent-Length: \d{2,4}\\r\\n\\r\\n/', $content);

        return $content;
    }

    /**
     * @depends testWriteHeader
     */
    public function testWriteContent($content)
    {
        $this->assertMatchesRegularExpression('/{"From":"\+15555555555","To":"\+16666666666","Body":"test1"}/', $content);
    }

    public function testWriteContentV1WithoutToAndFromNumbers()
    {
        $this->createHandler('token', 'auth_id', false, null, Logger::CRITICAL, true, true, 'twilio.foo.bar');
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'test1'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/{"From":false,"To":null,"Body":"test1"}/', $content);

        return $content;
    }

    /**
     * @depends testWriteCustomHostHeader
     */
    public function testWriteContentNotify($content)
    {
        $this->assertMatchesRegularExpression('/{"From":"\+15555555555","To":"\+16666666666","Body":"test1"}/', $content);
    }

    public function testWriteWithComplexMessage()
    {
        $this->createHandler();
        $this->handler->handle($this->getRecord(Logger::CRITICAL, 'Backup of database example finished in 16 minutes.'));
        fseek($this->res, 0);
        $content = fread($this->res, 1024);

        $this->assertMatchesRegularExpression('/{"From":"\+15555555555","To":"\+16666666666","Body":"Backup of database example finished in 16 minutes\."}/', $content);
    }

    private function createHandler($authToken = 'token', $authId = 'auth_id', $fromNumber = '+15555555555', $toNumber = '+16666666666', $level = Logger::CRITICAL, $bubble = true, $useSSL = true, $host = 'api.twilio.com', $version = TwilioHandler::API_V1)
    {
        $constructorArgs = array($authToken, $authId, $fromNumber, $toNumber, Logger::DEBUG, true, true, $host, $version);
        $this->res = fopen('php://memory', 'a');
        $this->handler = $this->getMockBuilder(TwilioHandler::class)
            ->setMethods(['fsockopen', 'streamSetTimeout', 'closeSocket'])
            ->setConstructorArgs($constructorArgs)
            ->getMock();

        $reflectionProperty = new \ReflectionProperty('\Monolog\Handler\SocketHandler', 'connectionString');
        $reflectionProperty->setAccessible(true);
        $reflectionProperty->setValue($this->handler, 'localhost:1234');

        $this->handler->expects($this->any())
            ->method('fsockopen')
            ->will($this->returnValue($this->res));
        $this->handler->expects($this->any())
            ->method('streamSetTimeout')
            ->will($this->returnValue(true));
        $this->handler->expects($this->any())
            ->method('closeSocket')
            ->will($this->returnValue(true));

        $this->handler->setFormatter($this->getIdentityFormatter());
    }
}


src/Handler/SMSHandler.php
<?php

namespace Tylercd100\Monolog\Handler;

use Exception;
use Monolog\Formatter\FormatterInterface;
use Monolog\Handler\SocketHandler;
use Monolog\Logger;
use Tylercd100\Monolog\Formatter\SMSFormatter;

abstract class SMSHandler extends SocketHandler
{

    /**
     * @var string
     */
    protected $authToken;

    /**
     * @var string
     */
    protected $authId;

    /**
     * @var string
     */
    protected $fromNumber;

    /**
     * @var string
     */
    protected $toNumber;

    /**
     * @var string
     */
    protected $host;

    /**
     * @var string
     */
    protected $version;

    /**
     * @var integer
     */
    protected $limit;

    /**
     * @param string $authToken  Plivo API Auth Token
     * @param string $authId     Plivo API Auth ID
     * @param string $fromNumber The phone number that will be shown as the sender ID
     * @param string $toNumber   The phone number to which the message will be sent
     * @param int    $level      The minimum logging level at which this handler will be triggered
     * @param bool   $bubble     Whether the messages that are handled can bubble up the stack or not
     * @param bool   $useSSL     Whether to connect via SSL.
     * @param string $host       The Plivo server hostname.
     * @param string $version    The Plivo API version (default PlivoHandler::API_V1)
     * @param int    $limit      The character limit
     */
    public function __construct($authToken, $authId, $fromNumber, $toNumber, $level = Logger::CRITICAL, $bubble = true, $useSSL = true, $host = 'api.plivo.com', $version = null, $limit = 160)
    {
        if (empty($version)) {
            throw new Exception('API Version is empty');
        }

        $connectionString = $useSSL ? 'ssl://'.$host.':443' : $host.':80';
        parent::__construct($connectionString, $level, $bubble);

        $this->authToken  = $authToken;
        $this->authId     = $authId;
        $this->fromNumber = $fromNumber;
        $this->toNumber   = $toNumber;
        $this->host       = $host;
        $this->version    = $version;
        $this->limit      = $limit;
    }

    /**
     * {@inheritdoc}
     *
     * @param  array  $record
     * @return string
     */
    protected function generateDataStream(array $record) :string
    {
        $content = $this->buildContent($record);
        return $this->buildHeader($content) . $content;
    }

    /**
     * Builds the body of API call
     *
     * @param  array  $record
     * @return string
     */
    abstract protected function buildContent($record);

    /**
     * Builds the URL for the API call
     *
     * @return string
     */
    abstract protected function buildRequestUrl();

    /**
     * Builds the header of the API call
     *
     * @param  string $content
     * @return string
     */
    private function buildHeader($content)
    {
        $auth = $this->authToken;

        if ($this->authId) {
            $auth = "Basic " . base64_encode($this->authId.":".$this->authToken);
        }

        $header = $this->buildRequestUrl();

        $header .= "Host: {$this->host}\r\n";
        $header .= "Authorization: ".$auth."\r\n";
        $header .= "Content-Type: application/json\r\n";
        $header .= "Content-Length: " . strlen($content) . "\r\n";
        $header .= "\r\n";
        return $header;
    }



    /**
     * {@inheritdoc}
     *
     * @param array $record
     */
    protected function write(array $record) :void
    {
        parent::write($record);
        $this->closeSocket();
    }

    /**
     * {@inheritdoc}
     */
    protected function getDefaultFormatter(): FormatterInterface
    {
        return new SMSFormatter();
    }
}


src/Handler/PlivoHandler.php
<?php

namespace Tylercd100\Monolog\Handler;

use Exception;
use Monolog\Logger;

/**
* Plivo - Monolog Handler
* @url https://www.plivo.com/docs/api/message/
*/
class PlivoHandler extends SMSHandler
{
    /**
     * API version 1
     */
    const API_V1 = 'v1';

    /**
     * @param string $authToken  Plivo API Auth Token
     * @param string $authId     Plivo API Auth ID
     * @param string $fromNumber The phone number that will be shown as the sender ID
     * @param string $toNumber   The phone number to which the message will be sent
     * @param int    $level      The minimum logging level at which this handler will be triggered
     * @param bool   $bubble     Whether the messages that are handled can bubble up the stack or not
     * @param bool   $useSSL     Whether to connect via SSL.
     * @param string $host       The Plivo server hostname.
     * @param string $version    The Plivo API version (default PlivoHandler::API_V1)
     * @param int    $limit      The character limit
     */
    public function __construct($authToken, $authId, $fromNumber, $toNumber, $level = Logger::CRITICAL, $bubble = true, $useSSL = true, $host = 'api.plivo.com', $version = self::API_V1, $limit = 160)
    {
        if ($version !== self::API_V1) {
            throw new Exception('API Version \'{$version}\' is not supported!');
        }
        parent::__construct($authToken, $authId, $fromNumber, $toNumber, $level, $bubble, $useSSL, $host, $version, $limit);
    }
    /**
     * {@inheritdoc}
     *
     * @param  array  $record
     * @return string
     */
    protected function buildContent($record)
    {
        if (strlen($record['formatted']) > $this->limit) {
            $record['formatted'] = substr($record['formatted'], 0, $this->limit);
        }

        $dataArray = array(
            'src'  => $this->fromNumber,
            'dst'  => $this->toNumber,
            'text' => $record['formatted'],
        );
        return json_encode($dataArray);
    }

    /**
     * Builds the URL for the API call
     *
     * @return string
     */
    protected function buildRequestUrl()
    {
        return "POST /{$this->version}/Account/{$this->authId}/Message/ HTTP/1.1\r\n";
    }
}


src/Handler/ClickatellHandler.php
<?php

namespace Tylercd100\Monolog\Handler;

use Exception;
use Monolog\Logger;

/**
* Clickatell - Monolog Handler
* @url https://www.clickatell.com/developers/api-documentation/rest-api-request-parameters
*/
class ClickatellHandler extends SMSHandler
{
    /**
     * API version 1
     */
    const API_V1 = '2010-04-01';

    /**
     * @param string $secret     Twilio API Secret Token
     * @param string $fromNumber The phone number that will be shown as the sender ID
     * @param string $toNumber   The phone number to which the message will be sent
     * @param int    $level      The minimum logging level at which this handler will be triggered
     * @param bool   $bubble     Whether the messages that are handled can bubble up the stack or not
     * @param bool   $useSSL     Whether to connect via SSL.
     * @param string $host       The Twilio server hostname.
     * @param string $version    The Twilio API version (default ClickatellHandler::API_V1)
     * @param int    $limit      The character limit
     */
    public function __construct($secret, $fromNumber, $toNumber, $level = Logger::CRITICAL, $bubble = true, $useSSL = true, $host = 'platform.clickatell.com', $version = self::API_V1, $limit = 160)
    {
        if ($version !== self::API_V1) {
            throw new Exception("API Version \'{$version}\' is not supported!");
        }
        parent::__construct($secret, null, $fromNumber, $toNumber, $level, $bubble, $useSSL, $host, $version, $limit);
    }

    /**
     * {@inheritdoc}
     *
     * @param  array  $record
     * @return string
     */
    protected function buildContent($record)
    {
        if (strlen($record['formatted']) > $this->limit) {
            $record['formatted'] = substr($record['formatted'], 0, $this->limit);
        }

        $dataArray = [
            'content' => $record['formatted'],
            'to' => (!is_array($this->toNumber)? [$this->toNumber] : $this->toNumber)
        ];

        ($this->fromNumber)? $dataArray['from'] = $this->fromNumber : false;

        return json_encode($dataArray);
    }

    /**
     * Builds the URL for the API call
     *
     * @return string
     */
    protected function buildRequestUrl()
    {
        return "POST /messages HTTP/1.1\r\n";
    }
}


src/Handler/TwilioHandler.php
<?php

namespace Tylercd100\Monolog\Handler;

use Exception;
use Monolog\Logger;

/**
* Twilio - Monolog Handler
* @url https://www.twilio.com/docs/api/rest/sending-messages
*/
class TwilioHandler extends SMSHandler
{
    /**
     * API version 1
     */
    const API_V1 = '2010-04-01';

    /**
     * @param string $secret     Twilio API Secret Token
     * @param string $sid        Twilio API SID
     * @param string $fromNumber The phone number that will be shown as the sender ID
     * @param string $toNumber   The phone number to which the message will be sent
     * @param int    $level      The minimum logging level at which this handler will be triggered
     * @param bool   $bubble     Whether the messages that are handled can bubble up the stack or not
     * @param bool   $useSSL     Whether to connect via SSL.
     * @param string $host       The Twilio server hostname.
     * @param string $version    The Twilio API version (default TwilioHandler::API_V1)
     * @param int    $limit      The character limit
     */
    public function __construct($secret, $sid, $fromNumber, $toNumber, $level = Logger::CRITICAL, $bubble = true, $useSSL = true, $host = 'api.twilio.com', $version = self::API_V1, $limit = 160)
    {
        if ($version !== self::API_V1) {
            throw new Exception('API Version \'{$version}\' is not supported!');
        }
        parent::__construct($secret, $sid, $fromNumber, $toNumber, $level, $bubble, $useSSL, $host, $version, $limit);
    }

    /**
     * {@inheritdoc}
     *
     * @param  array  $record
     * @return string
     */
    protected function buildContent($record)
    {
        if (strlen($record['formatted']) > $this->limit) {
            $record['formatted'] = substr($record['formatted'], 0, $this->limit);
        }

        $dataArray = array(
            'From' => $this->fromNumber,
            'To'   => $this->toNumber,
            'Body' => $record['formatted'],
        );
        return json_encode($dataArray);
    }

    /**
     * Builds the URL for the API call
     *
     * @return string
     */
    protected function buildRequestUrl()
    {
        return "POST /{$this->version}/Accounts/{$this->authId}/Messages.json HTTP/1.1\r\n";
    }
}
